addons/account/models/account_account.py

```diff
@@ -936,56 +950,72 @@ def name_create(self, name):
-                code_by_company_list.append(code_by_company)
-
-                vals['company_ids'] = [Command.set(companies[0].ids)]
-
-            check_company_vals_list = [{fname: vals.pop(fname) for fname in check_company_fields if fname in vals} for vals in vals_list_for_company]
-
-            new_accounts = super(AccountAccount, self.with_context({**self.env.context, 'allowed_company_ids': companies.ids})) \
-                                .create(vals_list_for_company)
-
-            # Add the other codes, companies and check_company fields on each account.
-            for new_account, code_by_company, check_company_vals in zip(new_accounts, code_by_company_list, check_company_vals_list):
-                for company_id, code in code_by_company.items():
-                    if company_id != companies[0].id:
-                        new_account.with_context({'allowed_company_ids': [company_id, companies[0].id]}).code = code
-                if len(companies) > 1:
-                    check_company_vals['company_ids'] = [Command.link(company.id) for company in companies[1:]]
-                if check_company_vals:
-                    new_account.write(check_company_vals)
+                if vals.get('code_mapping_ids') is False:
+                    vals['code_mapping_ids'] = []
+
+                if vals.get('code_mapping_ids'):
+                    # Prevent anything other than CREATE in `code_mapping_ids`
+                    vals['code_mapping_ids'] = [command for command in (vals['code_mapping_ids'] or []) if command[0] == Command.CREATE]
+                    code_by_company = {command[2].get('company_id'): command[2].get('code') for command in vals['code_mapping_ids']}
+
+                    # We can only have values either in `code` or in `code_mapping_ids`, otherwise `code` will get
+                    # written first and will then be protected, which will interfere with `code_mapping_ids.create`.
+                    # Therefore, if we have values in both, we put the value from `code` into `code_mapping_ids`.
+                    if 'code' in vals:
+                        code_in_active_company = vals.pop('code')
+                        if companies[0].id not in code_by_company:
+                            vals['code_mapping_ids'].append(Command.create({'company_id': companies[0].id, 'code': code_in_active_company}))
+                            code_by_company[companies[0].id] = code_in_active_company
+
+                    # Since we're giving values for `code_mapping_ids` and not `code`, `_compute_account_type` won't be called before
+                    # the record is created in DB, so the NOT NULL constraint on `account_type` will get triggered.
+                    # We therefore look here for a closest parent account in any of the companies to which the new account will belong,
+                    # and assign an appropriate account type.
+                    if 'account_type' not in vals:
```

Comment:
What about the tags?

Reply:
The tags are handled by the _compute_account_tags method